step-plans/step-plans-simple-yaml-parser/src/main/java/step/plans/simple/YamlPlanSerializer.java

```diff
@@ -0,0 +1,298 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with STEP.  If not, see <http://www.gnu.org/licenses/>.
+ ******************************************************************************/
+package step.plans.simple;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
+import org.bson.types.ObjectId;
+import org.reflections.Reflections;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import step.artefacts.handlers.JsonSchemaValidator;
+import step.core.Version;
+import step.core.accessors.AbstractIdentifiableObject;
+import step.core.accessors.DefaultJacksonMapperProvider;
+import step.core.artefacts.AbstractArtefact;
+import step.core.collections.Collection;
+import step.core.collections.CollectionFactory;
+import step.core.collections.Document;
+import step.core.collections.Filters;
+import step.core.collections.inmemory.InMemoryCollectionFactory;
+import step.core.dynamicbeans.DynamicValue;
+import step.core.plans.Plan;
+import step.core.scanner.CachedAnnotationScanner;
+import step.migration.MigrationManager;
+import step.plans.simple.deserializers.SimpleDynamicValueDeserializer;
+import step.plans.simple.deserializers.SimpleRootArtefactDeserializer;
+import step.plans.simple.migrations.AbstractSimplePlanMigrationTask;
+import step.plans.simple.model.SimpleRootArtefact;
+import step.plans.simple.model.SimpleYamlPlan;
+import step.plans.simple.model.SimpleYamlPlanVersions;
+import step.plans.simple.serializers.SimpleDynamicValueSerializer;
+import step.plans.simple.serializers.SimpleRootArtefactSerializer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.function.Supplier;
+
+import static step.core.scanner.Classes.newInstanceAs;
+
+public class YamlPlanSerializer {
+
+	private static final Logger log = LoggerFactory.getLogger(YamlPlanSerializer.class);
+
+	public static final String SIMPLE_PLANS_COLLECTION_NAME = "simplePlans";
+
+	private final ObjectMapper simpleYamlMapper;
+	private final ObjectMapper fullYamlMapper;
+
+	private final Supplier<ObjectId> idGenerator;
+	private Version currentVersion;
+	private String jsonSchema;
+	private final MigrationManager migrationManager;
+
+	// for tests
+	YamlPlanSerializer(Supplier<ObjectId> idGenerator, SimpleYamlPlanVersions.YamlPlanVersion currentVersion) {
+		if (currentVersion != null) {
+			setCurrentVersion(currentVersion);
+		} else {
+			// resolve version and schema automatically
```

Comment:
@jeromecomte @david-stephan 

I agree. This part is non-obvious, and maybe we can simplify it. Now in Step EE we have the `YamlPlanVersions` with `ACTUAL_VERSION` property. ACTUAL_VERSION defines the current version of Yaml Plan format and also the corrersponding json schema used to validate yaml plans.

But for Step EE we probably may want to have independent versioning (for example, when some new artefact is added to EE, but in OS nothing is changed we can increase the EE version, but keep the OS version untouched). And also we want to use the extended json-schema in EE project. 

Because in Step EE we use extension-classes, the version overriding was added to YamlPlanSerializerExtender (which initially was designed to add additional serialization/deserialization logic for EE). And because potentially in project it could be several YamlPlanSerializerExtenders to avoid confilcts I added the ExtendedYamlPlanVersion class with `priority` field to choose the extender with hightest priority. 

Now I see, that this prioritization was redundant, and we can just have the default ACTUAL_VERSION in OS and only one overriden version in EE (so I'll remove the redundant ExtendedYamlPlanVersion class).

Reply:
@jeromecomte @david-stephan 
But I'm still not sure about versioning strategy for EE project. Maybe it will be more understandable for users, if we always will have the same OS version and EE versions. 

In this case we can have the version defined in OS project, and in EE we only will use the extended (overridden) json-schema to validate yamls.

FYI: And anyway, for now in YAML plans we only have the 'version' property, so by yaml content we cannot determine if it was created for OS or for EE. But, maybe, it is not required in real-world use-cases...

Reply:
@iegorov777 After discussion with @david-stephan we would suggest to align the OS and EE versions. If a new version is required on EE, we'll simply increase it on OS, even if nothing changed on OS side.

Reply:
@jeromecomte 
I've aligned the json schema version in OS in EE. Now the actual version is defined in OS module and in EE we only can override the json schema (but not the version) via extensions.